File: src/services/riskCalculator.ts

// Define types for our flight data and risk assessment
export type FlightData = {
  speed: number;
  acceleration: number;
  temperature: number;
  humidity: number;
  windSpeed: number;
  visibility: number;
};

export type RiskAssessment = {
  score: number;
  level: 'low' | 'medium' | 'high';
  factors: {
    [key: string]: {
      value: number;
      risk: number;
      weight: number;
      description: string;
    }
  };
};

// Default thresholds for each parameter
const thresholds = {
  speed: {
    low: 500, // below this is low risk
    high: 600, // above this is high risk
    weight: 0.3, // importance in overall risk
    description: "Aircraft speed in knots"
  },
  acceleration: {
    low: -0.5, // stable or slight deceleration
    high: 2, // rapid acceleration
    weight: 0.1,
    description: "Rate of speed change"
  },
  temperature: {
    low: 5, // cold but not extreme
    high: 35, // hot conditions
    ideal: 15, // ideal temperature
    weight: 0.15,
    description: "Outside air temperature"
  },
  humidity: {
    low: 30,
    high: 80,
    weight: 0.05,
    description: "Air humidity percentage"
  },
  windSpeed: {
    low: 5,
    high: 25,
    weight: 0.25,
    description: "Current wind speed"
  },
  visibility: {
    low: 3, // below this is high risk
    high: 10, // above this is low risk
    weight: 0.15,
    description: "Visibility distance in km"
  }
};

// Calculate risk score for individual factor
function calculateFactorRisk(value: number, factorName: string): number {
  const threshold = thresholds[factorName as keyof typeof thresholds];
  
  // Special cases
  if (factorName === 'visibility') {
    // For visibility, higher is better (lower risk)
    if (value >= threshold.high) return 0;
    if (value <= threshold.low) return 1;
    return 1 - ((value - threshold.low) / (threshold.high - threshold.low));
  }
  
  if (factorName === 'temperature') {
    // Temperature risk is based on deviation from ideal
    const deviation = Math.abs(value - threshold.ideal);
    const range = Math.max(threshold.ideal - threshold.low, threshold.high - threshold.ideal);
    return Math.min(deviation / range, 1);
  }
  
  // Standard case - higher value means higher risk
  if (value <= threshold.low) return 0;
  if (value >= threshold.high) return 1;
  return (value - threshold.low) / (threshold.high - threshold.low);
}

// Calculate overall risk
export function calculateRisk(data: FlightData): RiskAssessment {
  // Initialize empty factors object
  const factors: RiskAssessment['factors'] = {};
  
  // Calculate risk for each factor
  let totalRisk = 0;
  let totalWeight = 0;
  
  for (const [key, value] of Object.entries(data)) {
    const threshold = thresholds[key as keyof typeof thresholds];
    if (!threshold) continue;
    
    const factorRisk = calculateFactorRisk(value, key);
    totalRisk += factorRisk * threshold.weight;
    totalWeight += threshold.weight;
    
    factors[key] = {
      value,
      risk: factorRisk,
      weight: threshold.weight,
      description: threshold.description
    };
  }
  
  // Normalize risk score to 0-100 range
  const normalizedScore = Math.round((totalRisk / totalWeight) * 100);
  
  // Determine risk level
  let level: 'low' | 'medium' | 'high';
  if (normalizedScore <= 30) {
    level = 'low';
  } else if (normalizedScore <= 70) {
    level = 'medium';
  } else {
    level = 'high';
  }
  
  return {
    score: normalizedScore,
    level,
    factors
  };
}

File: src/risk_calculator.py
import numpy as np
import pandas as pd
from typing import Dict, TypedDict, Literal
import skfuzzy as fuzz
from skfuzzy import control as ctrl

class FlightData(TypedDict):
    speed: float
    acceleration: float
    temperature: float
    humidity: float
    wind_speed: float
    visibility: float

class RiskAssessment(TypedDict):
    score: float
    level: Literal['low', 'medium', 'high']
    factors: Dict[str, Dict[str, float | str]]

# Define fuzzy membership functions
def create_fuzzy_system():
    # Universe of discourse
    speed_range = np.arange(0, 1000, 1)
    accel_range = np.arange(-5, 5, 0.1)
    temp_range = np.arange(-20, 50, 0.1)
    humidity_range = np.arange(0, 100, 1)
    wind_range = np.arange(0, 50, 1)
    visibility_range = np.arange(0, 20, 0.1)
    risk_range = np.arange(0, 1.1, 0.1)

    # Fuzzy membership functions
    speed = ctrl.Antecedent(speed_range, 'speed')
    acceleration = ctrl.Antecedent(accel_range, 'acceleration')
    temperature = ctrl.Antecedent(temp_range, 'temperature')
    humidity = ctrl.Antecedent(humidity_range, 'humidity')
    wind_speed = ctrl.Antecedent(wind_range, 'wind_speed')
    visibility = ctrl.Antecedent(visibility_range, 'visibility')
    risk = ctrl.Consequent(risk_range, 'risk')

    # Define membership functions
    speed['low'] = fuzz.trimf(speed_range, [0, 400, 500])
    speed['medium'] = fuzz.trimf(speed_range, [400, 550, 600])
    speed['high'] = fuzz.trimf(speed_range, [550, 600, 1000])

    acceleration['low'] = fuzz.trimf(accel_range, [-5, -0.5, 0])
    acceleration['medium'] = fuzz.trimf(accel_range, [-0.5, 0, 2])
    acceleration['high'] = fuzz.trimf(accel_range, [0, 2, 5])

    temperature['low'] = fuzz.trimf(temp_range, [-20, 0, 15])
    temperature['medium'] = fuzz.trimf(temp_range, [5, 15, 25])
    temperature['high'] = fuzz.trimf(temp_range, [15, 35, 50])

    humidity['low'] = fuzz.trimf(humidity_range, [0, 20, 40])
    humidity['medium'] = fuzz.trimf(humidity_range, [30, 50, 70])
    humidity['high'] = fuzz.trimf(humidity_range, [60, 80, 100])

    wind_speed['low'] = fuzz.trimf(wind_range, [0, 5, 15])
    wind_speed['medium'] = fuzz.trimf(wind_range, [10, 20, 30])
    wind_speed['high'] = fuzz.trimf(wind_range, [25, 35, 50])

    visibility['low'] = fuzz.trimf(visibility_range, [0, 2, 5])
    visibility['medium'] = fuzz.trimf(visibility_range, [3, 7, 10])
    visibility['high'] = fuzz.trimf(visibility_range, [8, 15, 20])

    risk['low'] = fuzz.trimf(risk_range, [0, 0.2, 0.4])
    risk['medium'] = fuzz.trimf(risk_range, [0.3, 0.5, 0.7])
    risk['high'] = fuzz.trimf(risk_range, [0.6, 0.8, 1.0])

    return {
        'speed': speed,
        'acceleration': acceleration,
        'temperature': temperature,
        'humidity': humidity,
        'wind_speed': wind_speed,
        'visibility': visibility,
        'risk': risk
    }

def calculate_risk(data: FlightData) -> RiskAssessment:
    fuzzy_system = create_fuzzy_system()
    factors = {}
    total_risk = 0
    weights = {
        'speed': 0.3,
        'acceleration': 0.1,
        'temperature': 0.15,
        'humidity': 0.05,
        'wind_speed': 0.25,
        'visibility': 0.15
    }

    # Calculate individual factor risks using fuzzy logic
    for factor, value in data.items():
        if factor in fuzzy_system:
            antecedent = fuzzy_system[factor]
            risk = fuzzy_system['risk']
            
            # Calculate membership values
            low_membership = fuzz.interp_membership(antecedent.universe, antecedent['low'].mf, value)
            medium_membership = fuzz.interp_membership(antecedent.universe, antecedent['medium'].mf, value)
            high_membership = fuzz.interp_membership(antecedent.universe, antecedent['high'].mf, value)
            
            # Calculate risk score using weighted average
            factor_risk = (low_membership * 0.2 + medium_membership * 0.5 + high_membership * 0.8)
            total_risk += factor_risk * weights[factor]
            
            factors[factor] = {
                'value': value,
                'risk': factor_risk,
                'weight': weights[factor],
                'description': get_factor_description(factor)
            }

    # Normalize risk score to 0-100 range
    normalized_score = round(total_risk * 100)
    
    # Determine risk level
    if normalized_score <= 30:
        level = 'low'
    elif normalized_score <= 70:
        level = 'medium'
    else:
        level = 'high'

    return {
        'score': normalized_score,
        'level': level,
        'factors': factors
    }

def get_factor_description(factor: str) -> str:
    descriptions = {
        'speed': 'Aircraft speed in knots',
        'acceleration': 'Rate of speed change',
        'temperature': 'Outside air temperature',
        'humidity': 'Air humidity percentage',
        'wind_speed': 'Current wind speed',
        'visibility': 'Visibility distance in km'
    }
    return descriptions.get(factor, '') 